### PR TITLE
Add support for lifetime dependence mangling

### PR DESCRIPTION
**Pull request by meg-gupta**

### CODE REVIEW

**Comment by meg-gupta** [comment]
@swift-ci test

**Comment by meg-gupta** [comment]
@swift-ci test

**Comment by meg-gupta** [comment]
@swift-ci test

File: lib/AST/ASTMangler.cpp [hunk 1]
```diff
@@ -2999,12 +3017,53 @@ void ASTMangler::appendTypeListElement(Identifier name, Type elementType,
...
+  // paramName) syntax, this can be a single case value check.
+  if (kind == LifetimeDependenceKind::Borrow ||
+      kind == LifetimeDependenceKind::Mutate) {
+    if (isSelfDependence) {
+      appendOperator("YLs");
+    } else {
+      appendOperator("Yls");
+    }
+  } else {
+    // If we converge on dependsOn(borrowed: paramName)/dependsOn(borrowed:
```
Comment (atrick) [inline lib/AST/ASTMangler.cpp:3043]: I think you meant `dependsOn(borrowed: paramName)/dependsOn(paramName`

**Review by atrick — COMMENTED** [review]
This all looks right to me. I'm not at all familiar with the mangler though. @eeckstein would you be able to give this an extra review?

I have one suggestion: eventually we will want to have dependencies between two parameters rather than on the result. Can you or @eeckstein, or @jckarter imagine how we would mangle a relative reference between two parameters? I don't think that would change how we mangle a dependence on the function result, but i just want to make sure it can be done in a straightforward, additive way.

**Comment by meg-gupta** [comment]
@atrick To represent lifetime dependence between parameters, I would imagine we will need a new mangling prefix followed by param indices or name references. This can be additive to the current mangling.

File: lib/AST/ASTMangler.cpp (hunk 1 above)
Comment (meg-gupta) [reply lib/AST/ASTMangler.cpp:3043]: Yes! If no other comments, I'll do this as a follow on.

**Review by eeckstein — COMMENTED** [review]
Implementation LGTM. Can you add some tests in test/Demangle/Inputs/manglings.txt?

**Comment by meg-gupta** [comment]
@swift-ci smoke test

**Pull request merged**